src/te/autodiff/jacobian.cc

```diff
@@ -0,0 +1,381 @@
+      } else if (op->name == "pow") {
+        auto x = op->args[0], y = op->args[1];
+        return expr * (Mutate(y)*log(x) + Mutate(x)*y/x);
+      } else if (op->name == "fabs") {
+        auto type = op->args[0].dtype();
+        return MulNode::make(Mutate(op->args[0]),
+                             SelectNode::make(GENode::make(op->args[0], make_zero(type)),
+                                              FloatImm(type, 1.0), FloatImm(type, -1.0)));
+      } else if (op->name == intrinsic::tvm_if_then_else) {
+        Array<PrimExpr> new_args = {op->args[0],
+                                    Mutate(op->args[1]),
+                                    Mutate(op->args[2])};
+        return CallNode::make(op->dtype, op->name, new_args,
+                              op->call_type, op->func, op->value_index);
+      } else if (piecewise_const.count(op->name)) {
+        return FloatImm(expr.dtype(), 0.0);
+      } else {
+        throw dmlc::Error("Derivative of this intrinsic is not implemented: " + op->name);
+      }
+    }
+    NOT_IMPLEMENTED
+  }
+
+  PrimExpr VisitExpr_(const AddNode* op) {
+    return AddNode::make(Mutate(op->a), Mutate(op->b));
+  }
+
+  PrimExpr VisitExpr_(const SubNode* op) {
+    return SubNode::make(Mutate(op->a), Mutate(op->b));
+  }
+
+  PrimExpr VisitExpr_(const MulNode* op) {
+    return AddNode::make(
+        MulNode::make(Mutate(op->a), op->b),
+        MulNode::make(op->a, Mutate(op->b)));
+  }
+
+  PrimExpr VisitExpr_(const DivNode* op) {
+    return DivNode::make(
+        SubNode::make(
+            MulNode::make(Mutate(op->a), op->b),
+            MulNode::make(op->a, Mutate(op->b))),
+        MulNode::make(op->b, op->b));
+  }
+
+  PrimExpr VisitExpr_(const ModNode* op) NOT_IMPLEMENTED
+
+  PrimExpr VisitExpr_(const FloorDivNode* op) {
+    return FloorDivNode::make(
+        SubNode::make(
+            MulNode::make(Mutate(op->a), op->b),
+            MulNode::make(op->a, Mutate(op->b))),
+        MulNode::make(op->b, op->b));
+  }
+
+  PrimExpr VisitExpr_(const FloorModNode* op) NOT_IMPLEMENTED
+
+  PrimExpr VisitExpr_(const MinNode* op) {
+    return SelectNode::make(LENode::make(op->a, op->b),
+        Mutate(op->a), Mutate(op->b));
+  }
+
+  PrimExpr VisitExpr_(const MaxNode* op) {
+    return SelectNode::make(GENode::make(op->a, op->b),
+        Mutate(op->a), Mutate(op->b));
+  }
+
+  PrimExpr VisitExpr_(const EQNode* op) NOT_IMPLEMENTED
+  PrimExpr VisitExpr_(const NENode* op) NOT_IMPLEMENTED
+  PrimExpr VisitExpr_(const LTNode* op) NOT_IMPLEMENTED
+  PrimExpr VisitExpr_(const LENode* op) NOT_IMPLEMENTED
+  PrimExpr VisitExpr_(const GTNode* op) NOT_IMPLEMENTED
+  PrimExpr VisitExpr_(const GENode* op) NOT_IMPLEMENTED
+  PrimExpr VisitExpr_(const AndNode* op) NOT_IMPLEMENTED
+  PrimExpr VisitExpr_(const OrNode* op) NOT_IMPLEMENTED
+
+  PrimExpr VisitExpr_(const ReduceNode* op) {
+    // This case is relatively difficult because a reduction expression
```

Comment:
Is it possible to have reduce inside other expression as well?

Reply:
seems to be a bit difficult. do you have an concrete example in mind?

Reply:
There are no concrete example. If you dont think it will happend then leave it as is.